pre_processing/read_era5_jasmin.F90

```diff
@@ -8,246 +8,339 @@
-   real(sreal)   :: ecmwf_lat(ecmwf%ydim)
-   integer(lint) :: pointer_x(preproc_dims%min_lon:preproc_dims%max_lon)
-   integer(lint) :: pointer_y(preproc_dims%min_lat:preproc_dims%max_lat)
-   real(sreal)   :: diff_lon(ecmwf%xdim), diff_lat(ecmwf%ydim)
+   type(ecmwf_t),    intent(inout)        :: ecmwf
+   type(preproc_nwp_fnames_t), intent(in) :: nwp_fnames
+   integer,          intent(in)           :: idx
+   integer,          intent(in)           :: nwp_flag
+
+   call ecmwf_wind_init(ecmwf)
+   call ecmwf_abvec_init(ecmwf)
+
+   ! loop over given files
+   call read_ecmwf_wind_nc_file(nwp_fnames%ci_f(idx), ecmwf)
+   call read_ecmwf_wind_nc_file(nwp_fnames%u10_f(idx), ecmwf)
+   call read_ecmwf_wind_nc_file(nwp_fnames%v10_f(idx), ecmwf)
+   call read_ecmwf_wind_nc_file(nwp_fnames%skt_f(idx), ecmwf)
+   call read_ecmwf_wind_nc_file(nwp_fnames%sd_f(idx), ecmwf)
+
+end subroutine read_era5_jasmin_wind_nc
+
+
+subroutine read_era5_jasmin_nc(nwp_fnames, idx, ecmwf, preproc_dims, preproc_geoloc, &
+     preproc_prtm, verbose, nwp_flag)
+
+    use orac_ncdf_m
+    use preproc_constants_m
+    use preproc_structures_m
+
+    implicit none
+
+    type(preproc_nwp_fnames_t), intent(in)    :: nwp_fnames
+    integer,                    intent(in)    :: idx
+    type(ecmwf_t),              intent(in)    :: ecmwf
+    type(preproc_dims_t),       intent(in)    :: preproc_dims
+    type(preproc_geoloc_t),     intent(in)    :: preproc_geoloc
+    type(preproc_prtm_t),       intent(inout) :: preproc_prtm
+    logical,                    intent(in)    :: verbose
+    integer,                    intent(in)    :: nwp_flag
+
+    integer(lint),     external            :: INTIN, INTOUT, INTF
+    integer(lint),     parameter           :: BUFFER = 2000000
+
+    integer(lint),            dimension(1) :: intv, old_grib, new_grib
+    real(dreal)                            :: grid(2), area(4)
+    character(len=20),        dimension(1) :: charv
+
+    real(sreal),       pointer             :: array2d(:,:), array3d(:,:,:)
+    integer(4)                             :: n, ni, nj, i, j, k, ivar
+    integer(4)                             :: fid, nvar
+    integer(4)                             :: old_len, new_len
+    character(len=20)                      :: name
+    logical                                :: three_d
+    real(sreal)   :: dummy2d(ecmwf%xdim,ecmwf%ydim)
+    real(sreal)   :: dummy3d(ecmwf%xdim,ecmwf%ydim,ecmwf%kdim,1)
+    real(sreal)   :: dummy3d_2(ecmwf%xdim,ecmwf%ydim,ecmwf%kdim)
+
+    ! open file
+    ! Do temperature on model levels
+    array3d => preproc_prtm%temperature
+    call ncdf_open(fid, nwp_fnames%t_f(idx), 'read_era5_jasmin_nc_t()')
+    call ncdf_read_array(fid, 't', dummy3d_2)
+    call preproc_3d_var(dummy3d_2, ecmwf, preproc_dims, preproc_geoloc, array3d)
+    if (verbose) print*, 'T)     Min: ', minval(array3d), ', Max: ', maxval(array3d)
+    call ncdf_close(fid, 'read_era5_jasmin_t()')
```

Comment:
I don't like having this many repeated blocks of code. Maybe a local convenience function that you pass a filename, the variable name, and the output variable? Also, the array pointer is unnecessary - you can just pass it to preproc_3d_var, right?